WIP: Go workspaces

thockin opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

thockin committed:
DO NOT MERGE: Make gengo/v2 in vendor, to hack

This only copies the files and fixes up imports, go.mod, and
vendor/modules.txt.

This depends on using gotip with support for workspaces+vendor.

TO REPEAT:
  pushd vendor/k8s.io
  rm -rf gengo/v2
  cp -a gengo gengo2
  mv gengo2 gengo/v2
  rm gengo/v2/LICENSE
  find gengo/v2 -type f | xargs grep -l 'k8s.io/gengo' | while read F; do sed -i 's|k8s.io/gengo|k8s.io/gengo/v2|g' $F; done
  git add gengo/v2
  popd

thockin committed:
codegen: prerelease-lifecycle on gengo/v2

./hack/make-rules/build.sh ./staging/src/k8s.io/code-generator/cmd/prerelease-lifecycle-gen/

thockin committed:
codegen: openapi on gengo/v2

./hack/make-rules/build.sh kube-openapi-hack/cmd/openapi-gen/

thockin committed:
Remove PRJ_SRC_PATH last users

thockin committed:
Workspaces: excise more vendor from golang.sh

```
$ make
go version go1.20.4 linux/amd64
+++ [0506 13:21:33] Building go targets for linux/amd64
    ./cmd/kube-proxy (static)
    ./cmd/kube-apiserver (static)
    ./cmd/kube-controller-manager (static)
    ./cmd/kubelet (non-static)
    ./cmd/kubeadm (static)
    ./cmd/kube-scheduler (static)
    ./staging/src/k8s.io/component-base/logs/kube-log-runner (static)
    ./staging/src/k8s.io/kube-aggregator (static)
    ./staging/src/k8s.io/apiextensions-apiserver (static)
    ./cluster/gce/gci/mounter (non-static)
    ./cmd/kubectl (static)
    ./cmd/kubectl-convert (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    ./test/e2e/e2e.test (test)
    ./test/conformance/image/go-runner (non-static)
    ./cmd/kubemark (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    ./test/e2e_node/e2e_node.test (test)
```

before:
real	4m31.100s
user	21m1.466s
sys	2m23.546s

after:
real	3m46.591s
user	16m34.665s
sys	2m1.553s

thockin committed:
Remove vendor from ldflags

thockin committed:
Workspaces: `make test` works

```
$ make test WHAT="./pkg/proxy/iptables ./staging/src/k8s.io/api"
+++ [0306 11:16:59] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/iptables	4.482s
ok  	k8s.io/api	9.234s
```

thockin committed:
Fix last run-in-gopath user: typecheck

This now works across modules

Verified by introducing errors in 2 different modules and running.

thockin committed:
Remove run-in-gopath.sh

thockin committed:
Clean up test.sh and get rid of GOPATH

This also allows slightly sloppier input, just as build does.

before:
```
$ ./hack/make-rules/test.sh pkg/proxy/util/
!!! [0521 16:01:29] specified test path '${GOPATH}/src/pkg/proxy/util/' does not exist
try changing "pkg/proxy/util/" to "./pkg/proxy/util/"

$ ./hack/make-rules/test.sh ./pkg/proxy/util/
+++ [0521 16:01:34] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/util	(cached)

$ ./hack/make-rules/test.sh k8s.io/kubernetes/pkg/proxy/util/
+++ [0521 16:01:39] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/util	(cached)

```

after:
```
$ ./hack/make-rules/test.sh pkg/proxy/util/
+++ [0521 15:58:45] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/util	1.906s

$ ./hack/make-rules/test.sh ./pkg/proxy/util/
+++ [0521 15:58:56] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/util	(cached)

$ ./hack/make-rules/test.sh k8s.io/kubernetes/pkg/proxy/util/
+++ [0521 15:59:05] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/proxy/util	(cached)
```

thockin committed:
Remove references to GOPATH

thockin committed:
Makefile: abort if symlinks in CWD

This does not have USER impact but it may have developer impact.  If the
path to you git repo includes a symlink, it will now ask you to change
to the non-symlink directory.

```
$ make
Makefile:69: The current directory appears to have symlinks, which confuses Go.  Try:
Makefile:70: `    cd /home/thockin/src/go/src/k8s.io/kubernetes  # cd $(pwd -P)
Makefile:71: *** Aborting.  Stop.
```

thockin committed:
Makefile: enable warn-undefined-variables

thockin committed:
Fix update/verify-yamlfmt.sh

thockin committed:
run update-yamlfmt

thockin committed:
Fix update/verify-mocks.sh

Use `git worktree`.

thockin committed:
Remove vendored symlinks to staging

This breaks all sorts of tooling, but emphasizes the things we need to
fix.

thockin committed:
Change the build_binaries path to use modules

This makes "new" and "old" setup_env functions. In subsequent commits,
all callers of the "old" form will be fixed, and the "new" will be
renamed back.

The old and new functions diff:

```diff
--- /tmp/old	2023-05-06 12:03:23.913951720 -0700
+++ /tmp/new	2023-05-06 12:06:37.688300514 -0700
@@ -17,9 +16,9 @@
   # Even in module mode, we need to set GOPATH for `go build` and `go install`
   # to work.  We build various tools (usually via `go install`) from a lot of
   # scripts.
-  #   * We can't set GOBIN because that does not work on cross-compiles.
-  #   * We could use `go build -o <something>`, but it's subtle when it comes
-  #     to cross-compiles and whether the <something> is a file or a directory,
+  #   * We can't just set GOBIN because that does not work on cross-compiles.
+  #   * We could always use `go build -o <something>`, but it's subtle wrt
+  #     cross-compiles and whether the <something> is a file or a directory,
   #     and EVERY caller has to get it *just* right.
   #   * We could leave GOPATH alone and let `go install` write binaries
   #     wherever the user's GOPATH says (or doesn't say).
@@ -30,7 +29,6 @@
   #
   # Eventually, when we no longer rely on run-in-gopath.sh we may be able to
   # simplify this some.
-  kube::golang::old::create_gopath_tree
   export GOPATH="${KUBE_GOPATH}"

   # If these are not set, set them now.  This ensures that any subsequent
@@ -41,23 +39,12 @@
   # Make sure our own Go binaries are in PATH.
   export PATH="${KUBE_GOPATH}/bin:${PATH}"

-  # Change directories so that we are within the GOPATH.  Some tools get really
-  # upset if this is not true.  We use a whole fake GOPATH here to collect the
-  # resultant binaries.
-  local subdir
-  subdir=$(kube::realpath . | sed "s|${KUBE_ROOT}||")
-  cd "${KUBE_GOPATH}/src/${KUBE_GO_PACKAGE}/${subdir}" || return 1
-
-  # Set GOROOT so binaries that parse code can work properly.
-  GOROOT=$(go env GOROOT)
-  export GOROOT
-
   # Unset GOBIN in case it already exists in the current session.
   # Cross-compiles will not work with it set.
   unset GOBIN

-  # This seems to matter to some tools
-  export GO15VENDOREXPERIMENT=1
+  # Explicitly turn on modules.
+  export GO111MODULE=on

   # GOMAXPROCS by default does not reflect the number of cpu(s) available
   # when running in a container, please see https://github.com/golang/go/issues/33803
@@ -68,7 +55,6 @@
     # shellcheck disable=SC2164
     popd >/dev/null
   fi
-
   GOMAXPROCS=${GOMAXPROCS:-$(ncpu)}
-  kube::log::status "Setting GOMAXPROCS: ${GOMAXPROCS}"
+  V=3 kube::log::status "Setting GOMAXPROCS: ${GOMAXPROCS}"
 }
```

Result: `make` works in some modes, across modules, but not all.

```
$ make WHAT="./cmd/kube-proxy/ k8s.io/kube-aggregator/pkg/cmd/server"
go version go1.20.4 linux/amd64
+++ [0506 12:52:55] Building go targets for linux/amd64
    k8s.io/kubernetes/./cmd/kube-proxy/ (non-static)
    k8s.io/kube-aggregator/pkg/cmd/server (non-static)

$ make WHAT="k8s.io/kubernetes/cmd/kube-proxy k8s.io/kube-aggregator/pkg/cmd/server"
go version go1.20.4 linux/amd64
+++ [0506 12:53:20] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kube-proxy (static)
    k8s.io/kube-aggregator/pkg/cmd/server (non-static)
```

but:

```
$ make WHAT="./staging/src/k8s.io/kube-aggregator/pkg/cmd/server"
go version go1.20.4 linux/amd64
+++ [0506 12:54:32] Building go targets for linux/amd64
    k8s.io/kubernetes/./staging/src/k8s.io/kube-aggregator/pkg/cmd/server (non-static)
no required module provides package k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/cmd/server; to add it:
	go get k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/cmd/server
!!! [0506 12:54:32] Call tree:
!!! [0506 12:54:32]  1: /home/thockin/src/go/src/k8s.io/kubernetes/hack/lib/golang.sh:845 kube::golang::build_some_binaries(...)
!!! [0506 12:54:32]  2: /home/thockin/src/go/src/k8s.io/kubernetes/hack/lib/golang.sh:1007 kube::golang::build_binaries_for_platform(...)
!!! [0506 12:54:32]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0506 12:54:32] Call tree:
!!! [0506 12:54:32]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0506 12:54:32] Call tree:
!!! [0506 12:54:32]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1
```

`go build` works:
```
$ go build ./cmd/kube-proxy ./staging/src/k8s.io/kube-aggregator/pkg/cmd/server

$ go build k8s.io/kubernetes/cmd/kube-proxy k8s.io/kube-aggregator/pkg/cmd/server
```

thockin committed:
FIXME: gengo-cleanup-notes

thockin committed:
Re-enable GOPROXY for build

Workspaces does not support vendoring, so this is just incompatible.

Result: `make` and `make test` work.  `make WHAT=...` does not always
work:

```
$ make WHAT="./staging/src/k8s.io/kube-aggregator/pkg/cmd/server"
go version go1.20.4 linux/amd64
+++ [0506 12:39:03] Setting GOMAXPROCS: 6
+++ [0506 12:39:04] Building go targets for linux/amd64
    k8s.io/kubernetes/./staging/src/k8s.io/kube-aggregator/pkg/cmd/server (non-static)
no required module provides package k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/cmd/server; to add it:
	go get k8s.io/kubernetes/staging/src/k8s.io/kube-aggregator/pkg/cmd/server
!!! [0506 12:39:04] Call tree:
!!! [0506 12:39:04]  1: /home/thockin/src/go/src/k8s.io/kubernetes/hack/lib/golang.sh:782 kube::golang::build_some_binaries(...)
!!! [0506 12:39:04]  2: /home/thockin/src/go/src/k8s.io/kubernetes/hack/lib/golang.sh:941 kube::golang::build_binaries_for_platform(...)
!!! [0506 12:39:04]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0506 12:39:04] Call tree:
!!! [0506 12:39:04]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0506 12:39:04] Call tree:
!!! [0506 12:39:04]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1
```

FIXME: If we pre-download and commit a full module cache, then we can
leave this set as insurance?

thockin committed:
FIXME: add FIXME (remove this commit)

thockin committed:
Excise KUBE_GOPATH from coverage build

thockin committed:
Remove old build code

thockin committed:
Rename KUBE_OUTPUT_BINPATH -> KUBE_OUTPUT_BIN

"...PATH" doesn't add anything and confuses it with KUBE_OUTPUT_SUBPATH.

thockin committed:
Retool, clarify, document KUBE_OUTPUT_* variables

thockin committed:
------------------------- PREFACTORING BEFORE THIS

This commit is just a marker in the history and must be removed before
merge.

thockin committed:
WIP: gengo/v2: main changes to make codegens run

* Lots of FIXMEs
* More to minimize
* Be clearer about package names and on-disk paths

thockin committed:
codegen: deepcopy on gengo/v2

Building on all the gengo changes and script changes, deepcopy-gen now
works in modules + workspaces mode.

build:
    ./hack/make-rules/build.sh ./staging/src/k8s.io/code-generator/cmd/deepcopy-gen/

run:
    _output/bin/deepcopy-gen --v=5 --logtostderr -i ./staging/src/k8s.io/api/node/v1 --bounding-dirs k8s.io/kubernetes,k8s.io/api -O zz_generated.deepcopy

debug:
    dlv exec _output/bin/deepcopy-gen -- --v=5 --logtostderr -i ./staging/src/k8s.io/api/node/v1 --bounding-dirs k8s.io/kubernetes,k8s.io/api -O zz_generated.deepcopy

thockin committed:
codegen: defaulter on gengo/v2

./hack/make-rules/build.sh ./staging/src/k8s.io/code-generator/cmd/defaulter-gen/

Status: defaulter runs and doesn't need "extra peer dirs" anymore

thockin committed:
codegen: conversion on gengo/v2

./hack/make-rules/build.sh ./staging/src/k8s.io/code-generator/cmd/conversion-gen/

Status: conversion runs and doesn't need "extra dirs" anymore

thockin committed:
gengo/v2: FIXME: gengo-cleanup-notes

thockin committed:
gengo/v2: gengo cleanup: remove commented code

thockin committed:
gengo/v2: simpler lookup

thockin committed:
gengo/v2: move a method near the only user of it

thockin committed:
gengo/v2: gengo2: remove unused code

thockin committed:
gengo/v2: gengo cleanup: move lock into struct

thockin committed:
gengo/v2: gengo cleanup: remove context

thockin committed:
gengo/v2: gengo: add FIXMEs

thockin committed:
gengo/v2: tighter API between args and builder

thockin committed:
gengo/v2: gengo cleanup: Remove FIXME

thockin committed:
------------------------------------ ADD TODO FILE

This commit just adds my TODO notes, and must be removed before merge.

thockin committed:
Fix verify-api-groups.sh

thockin committed:
Fix verify-cli-conventions.sh

thockin committed:
WIP: codegen lister-gen

hack/update-codegen.sh makes no diffs

aojea commented:
/cc

aojea commented on pkg/util/coverage/fake_test_deps.go:
```diff
@@ -61,3 +63,27 @@ func (fakeTestDeps) WriteHeapProfile(io.Writer) error {
 func (fakeTestDeps) WriteProfileTo(string, io.Writer, int) error {
 	return nil
 }
+
+//nolint:unused // U1000 see comment above, we know it's unused normally.
+func (fakeTestDeps) CheckCorpus([]interface{}, []reflect.Type) error {
+	return nil
+}
+
+//nolint:unused // U1000 see comment above, we know it's unused normally.
+func (fakeTestDeps) SetPanicOnExit0(bool) {}
+
+//nolint:unused // U1000 see comment above, we know it's unused normally.
+func (fakeTestDeps) CoordinateFuzzing(time.Duration, int64, time.Duration, int64, int, []corpusEntry, []reflect.Type, string, string) error {
+	return nil
+}
+
+//nolint:unused // U1000 see comment above, we know it's unused normally.
+func (fakeTestDeps) RunFuzzWorker(func(corpusEntry) error) error {
+	return nil
+}
+
+//nolint:unused // U1000 see comment above, we know it's unused normally.
+func (fakeTestDeps) ResetCoverage() {}
+
+//nolint:unused // U1000 see comment above, we know it's unused normally.
+func (fakeTestDeps) SnapshotCoverage() {}
```
https://github.com/google/gvisor/blob/0968254ce715c0d8e6457558d55cb405a545ad2f/test/runtimes/runner/lib/lib.go#L212-L241 ??

aojea replied:
it seems you have to copy `corpusEntry` 

```
// Copied from testing/fuzz.go.
type corpusEntry = struct {
	Parent     string
	Path       string
	Data       []byte
	Values     []interface{}
	Generation int
	IsSeed     bool
}
```

pohly commented:
/cc

I want to investigate a bit how hack/pin-dependency.sh and hack/update-vendor.sh could be adapted.

One idea that I had was to emulate the current, non-workspace setup and then run "go mod vendor". The result then could be checked in, which would address the concern about reviewing dependency changes. It would not address the concern about being able to build without downloading sources.

jhoblitt commented:
> One idea that I had was to emulate the current, non-workspace setup and then run "go mod vendor". The result then could be checked in, which would address the concern about reviewing dependency changes. It would not address the concern about being able to build without downloading sources.

It seems important to decide if offline builds are a hard requirement. If it is, I wonder if a hacked up local go mod proxy would be workable.

aojea commented on cmd/yamlfmt/yamlfmt.go:
```diff
@@ -32,19 +33,23 @@ func main() {
+				fmt.Fprintf(os.Stderr, "%s: %v\n", path, err)
+				continue
 			}
 		}
```
TODO, os.exit(1) if err != nil

aojea commented on hack/make-rules/helpers/go2make/go2make.go:
```diff
@@ -185,45 +184,41 @@ func (emit emitter) loadPackages(targets ...string) ([]*packages.Package, error)
-				// ignore errors like "build constraints exclude all Go files"
-				debug("    ignoring error")
-				continue
-			}
```
this error was ignored ALWAYS, don't we need to keep ignoring it?

fedebongio commented:
/triage accepted

pohly commented on go.work:
```diff
@@ -0,0 +1,725 @@
+	./staging/src/k8s.io/sample-controller
+)
+
+replace bazil.org/fuse => bazil.org/fuse v0.0.0-20160811212531-371fbbdaa898
```
Why do we need all these `replace` statements?

My understanding is that without them, Go will use the `replace` statements from the different `go.mod` files, so we shouldn't need to replicate them.

As an experiment, I ripped them all out. "make all" still worked, after resolving one problem: `staging/src/k8s.io/code-generator/examples` is not consistent with the other packages, probably because update-vendor.sh ignores it (?). We probably should fix that or, if that doesn't work, not include it in go.work file.

 However, after "make clean" the same command failed:
 ```
+++ [0608 10:39:54] Building go targets for linux/amd64
    ./staging/src/k8s.io/code-generator/cmd/deepcopy-gen (non-static)
make[1]: *** No rule to make target '.make/deps/by-path/./staging/src/k8s.io/code-generator/examples/HyphenGroup/apis/example/v1/_pkg', needed by 'staging/src/k8s.io/code-generator/examples/HyphenGroup/apis/example/v1/zz_generated.deepcopy.go'.  Stop.
make: *** [Makefile:536: generated_files] Error 2
```

So the replace statements are needed for something, it's just not entirely clear to me for what.

pohly replied:
It worked without any `replace` statements after re-adding `./staging/src/k8s.io/code-generator/examples` and fixing its `go.mod` file.

pohly replied:
See https://github.com/thockin/kubernetes/pull/12

pohly committed:
Generate go.work files

This creates go.work and enables Go Workspaces.  This is a file that
includes info on all Go modules in k/k.

Make this part of the normal update/verify sequence.

The top-level go.work file contains no replace statements. Instead, the
replace statements in the individual go.mod files are used. For this to
work, replace statements in the individual go.mod files have to be
consistent.

hack/tools has different dependencies and can't be in the main
workspace, so this adds a go.work just for that.  Without this, go tries
to consider all deps in all modules and pick one that works for all.
This is problematic because there are so many of them that it is
difficult to manage.

For example k/k depends on an older version of a lib that gloangci-lint
needs (transitively) and it breaks.

Result: `go` commands work across modules:

```
$ go list ./pkg/proxy/iptables/ ./staging/src/k8s.io/api/core/v1/
k8s.io/kubernetes/pkg/proxy/iptables
k8s.io/api/core/v1

$ go build ./pkg/proxy/iptables/ ./staging/src/k8s.io/api

$ go test ./pkg/proxy/iptables/ ./staging/src/k8s.io/api
ok  	k8s.io/kubernetes/pkg/proxy/iptables	0.360s
ok  	k8s.io/api	2.302s
```

Result: `make` fails:

```
$ make
go version go1.20.4 linux/amd64
+++ [0506 11:13:36] Setting GOMAXPROCS: 6
+++ [0506 11:13:37] Building go targets for linux/amd64
    k8s.io/kubernetes/cmd/kube-proxy (static)
    k8s.io/kubernetes/cmd/kube-apiserver (static)
    k8s.io/kubernetes/cmd/kube-controller-manager (static)
    k8s.io/kubernetes/cmd/kubelet (non-static)
    k8s.io/kubernetes/cmd/kubeadm (static)
    k8s.io/kubernetes/cmd/kube-scheduler (static)
    k8s.io/component-base/logs/kube-log-runner (static)
    k8s.io/kube-aggregator (static)
    k8s.io/apiextensions-apiserver (static)
    k8s.io/kubernetes/cluster/gce/gci/mounter (non-static)
    k8s.io/kubernetes/cmd/kubectl (static)
    k8s.io/kubernetes/cmd/kubectl-convert (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    k8s.io/kubernetes/test/e2e/e2e.test (test)
    k8s.io/kubernetes/test/conformance/image/go-runner (non-static)
    k8s.io/kubernetes/cmd/kubemark (static)
    github.com/onsi/ginkgo/v2/ginkgo (non-static)
    k8s.io/kubernetes/test/e2e_node/e2e_node.test (test)
cmd/kube-proxy/app/server_others.go:32:2: module lookup disabled by GOPROXY=off
cmd/kube-proxy/app/server_others.go:33:2: module lookup disabled by GOPROXY=off
pkg/util/ipvs/ipvs_linux.go:30:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/util/cgroups_linux.go:22:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/util/cgroups_linux.go:23:2: module lookup disabled by GOPROXY=off
pkg/proxy/ipvs/netlink_linux.go:31:2: module lookup disabled by GOPROXY=off
pkg/volume/util/selinux.go:22:2: module lookup disabled by GOPROXY=off
pkg/volume/util/selinux.go:23:2: module lookup disabled by GOPROXY=off
pkg/apis/batch/validation/validation.go:25:2: module lookup disabled by GOPROXY=off
pkg/volume/csi/csi_client.go:28:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/qos_container_manager_linux.go:31:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:35:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:36:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:31:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:32:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:33:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:37:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/helpers_linux.go:25:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:38:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:39:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_linux.go:40:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_cloudproviders.go:25:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cadvisor/cadvisor_cloudproviders.go:26:2: module lookup disabled by GOPROXY=off
pkg/volume/portworx/portworx_util.go:25:2: module lookup disabled by GOPROXY=off
pkg/volume/portworx/portworx_util.go:26:2: module lookup disabled by GOPROXY=off
pkg/volume/portworx/portworx.go:29:2: module lookup disabled by GOPROXY=off
pkg/volume/portworx/portworx_util.go:28:2: module lookup disabled by GOPROXY=off
pkg/volume/portworx/portworx_util.go:29:2: module lookup disabled by GOPROXY=off
pkg/security/apparmor/validate.go:24:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/cgroup_manager_linux.go:31:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/cgroup_manager_linux.go:32:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/cgroup_manager_linux.go:33:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/cgroup_manager_linux.go:34:2: module lookup disabled by GOPROXY=off
pkg/kubelet/cm/container_manager_linux.go:39:2: module lookup disabled by GOPROXY=off
cmd/kubeadm/app/phases/addons/dns/dns.go:25:2: module lookup disabled by GOPROXY=off
cmd/kubeadm/app/preflight/checks.go:45:2: module lookup disabled by GOPROXY=off
pkg/kubelet/kuberuntime/kuberuntime_container.go:41:2: module lookup disabled by GOPROXY=off
pkg/kubelet/kubelet_server_journal.go:37:2: module lookup disabled by GOPROXY=off
pkg/kubelet/nodeshutdown/systemd/inhibit_linux.go:29:2: module lookup disabled by GOPROXY=off
cmd/kubelet/app/options/globalflags_linux.go:29:2: module lookup disabled by GOPROXY=off
cmd/kubelet/app/options/globalflags_linux.go:30:2: module lookup disabled by GOPROXY=off
cmd/kubelet/app/options/globalflags_linux.go:31:2: module lookup disabled by GOPROXY=off
pkg/kubelet/server/server.go:39:2: module lookup disabled by GOPROXY=off
cmd/kubelet/app/options/globalflags_linux.go:34:2: module lookup disabled by GOPROXY=off
pkg/kubelet/oom/oom_watcher_linux.go:30:2: module lookup disabled by GOPROXY=off
pkg/kubelet/server/server.go:40:2: module lookup disabled by GOPROXY=off
!!! [0506 11:13:39] Call tree:
!!! [0506 11:13:39]  1: /home/thockin/src/go/src/k8s.io/kubernetes/hack/lib/golang.sh:771 kube::golang::build_some_binaries(...)
!!! [0506 11:13:39]  2: /home/thockin/src/go/src/k8s.io/kubernetes/hack/lib/golang.sh:941 kube::golang::build_binaries_for_platform(...)
!!! [0506 11:13:39]  3: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0506 11:13:39] Call tree:
!!! [0506 11:13:39]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [0506 11:13:39] Call tree:
!!! [0506 11:13:39]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make: *** [Makefile:92: all] Error 1
```

TO REPEAT: ./hack/update-go-workspace.sh

pohly commented on hack/lib/golang.sh:
```diff
@@ -552,6 +554,54 @@ kube::golang::setup_env() {
+  # simplify this some.
+  export GOPATH="${KUBE_GOPATH}"
+
+  # Keep caches local.
```
Why? Is that part of the experiment to see whether the GOCACHE might be checked into git?

pohly commented:
> 



> > One idea that I had was to emulate the current, non-workspace setup and then run "go mod vendor". The result then could be checked in, which would address the concern about reviewing dependency changes. It would not address the concern about being able to build without downloading sources.
> 
> It seems important to decide if offline builds are a hard requirement. If it is, I wonder if a hacked up local go mod proxy would be workable.

The problem with a "proper" Go mod proxy is that we would have to provide the full source code that matches the checksums in go.sum. I found a way around that: when the module download directory contains a .zip file and alongside it a checksum file, then Go trusts that file and compares it against the expected go.sum file. The trick in https://github.com/thockin/kubernetes/pull/13 is to generate a .zip file from the source in vendor *and* a checksum file that matches the go.sum file. Then when Go looks for the source, it uses the .zip file content without checking it again.

Another trick is that we must store the original, unmodified go.mod files under vendor. They are needed for the module cache.

I also tried to populate the unpacked sources directly in the module cache, but couldn't get it to work. Even if it had worked, creating just the zip archives seems simpler. I must point out that it makes "zip" a required tool. That probably wasn't the case before. It's not guaranteed to be installed on all Linux systems.

If we go with this, then we would continue to use pin-dependency.sh and update-vendor.sh. The latter can be simplified because we don't need any replace directives in staging repos because we never build them in non-workspace. I haven't done that yet because I want feedback first and update-vendor.sh is currently not working because of the ginkgo/v2 hack.

jhoblitt commented:
I think that requiring `zip` to be provided by the dev env would not be an onerous requirement on top of `go`, `make`, `sh`, etc..

thockin committed:
WIP: codegen subdirs are broken

thockin committed:
WIP: fixing openapi file splats

thockin committed:
WIP: change gengo args

thockin committed:
WIP: gengo/v2: FormatOnly on imports

Running goimports in module mode is slower than before.  Running it with
workspaces is even slower.

This change DRAMATICALLY improves code-generation time (back to where it
was before modules).

TODO: This seems to need other changes if I apply it in prefactoring,
but not here.  Not sure why.  It emits imports for "this package", which
goimports used to fix.

We can use the "local" capability in the ImportTracker which prevents
"this package" from being tracked, which causes generated code to use
different import names (e.g. "v1" is available now where it used to be
consumed by "this package").

diff --git a/staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/packages.go b/staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/packages.go
index bfeffda593d..66d195e475a 100644
--- a/staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/packages.go
+++ b/staging/src/k8s.io/code-generator/cmd/applyconfiguration-gen/generators/packages.go
@@ -177,7 +177,7 @@ func generatorForApplyConfigurationsPackage(outputPackagePath string, boilerplat
 					localPackage:  packageName,
 					groupVersion:  gv,
 					applyConfig:   toGenerate,
-					imports:       generator.NewImportTracker(),
+					imports:       generator.NewImportTracker(packageName.Package),
 					refGraph:      refs,
 					openAPIType:   openAPIType,
 				})
@@ -198,7 +198,7 @@ func generatorForUtils(outPackagePath string, boilerplate []byte, groupVersions
 					OptionalName: "utils",
 				},
 				outputPackage:        outPackagePath,
-				imports:              generator.NewImportTracker(),
+				imports:              generator.NewImportTracker(outPackagePath),
 				groupVersions:        groupVersions,
 				typesForGroupVersion: applyConfigsForGroupVersion,
 				groupGoNames:         groupGoNames,
@@ -219,7 +219,7 @@ func generatorForInternal(outPackagePath string, boilerplate []byte, models *typ
 					OptionalName: "internal",
 				},
 				outputPackage: outPackagePath,
-				imports:       generator.NewImportTracker(),
+				imports:       generator.NewImportTracker(outPackagePath),
 				typeModels:    models,
 			})
 			return generators
diff --git a/vendor/k8s.io/gengo/generator/import_tracker.go b/vendor/k8s.io/gengo/generator/import_tracker.go
index 60c899ac464..408e91f7a39 100644
--- a/vendor/k8s.io/gengo/generator/import_tracker.go
+++ b/vendor/k8s.io/gengo/generator/import_tracker.go
@@ -26,8 +26,8 @@ import (
 	"k8s.io/gengo/types"
 )

-func NewImportTracker(typesToAdd ...*types.Type) namer.ImportTracker {
-	tracker := namer.NewDefaultImportTracker(types.Name{})
+func NewImportTracker(thisPkg string, typesToAdd ...*types.Type) namer.ImportTracker {
+	tracker := namer.NewDefaultImportTracker(thisPkg)
 	tracker.IsInvalidType = func(*types.Type) bool { return false }
 	tracker.LocalName = func(name types.Name) string { return golangTrackerLocalName(&tracker, name) }
 	tracker.PrintImport = func(path, name string) string { return name + " \"" + path + "\"" }
diff --git a/vendor/k8s.io/gengo/namer/import_tracker.go b/vendor/k8s.io/gengo/namer/import_tracker.go
index 37094b2deb5..9f24f6caeea 100644
--- a/vendor/k8s.io/gengo/namer/import_tracker.go
+++ b/vendor/k8s.io/gengo/namer/import_tracker.go
@@ -32,7 +32,7 @@ type DefaultImportTracker struct {
 	// there is code, but "go" is not a legal name for a package, so we put
 	// it here to prevent us from naming any package "go")
 	nameToPath map[string]string
-	local      types.Name
+	thisPkg    string

 	// Returns true if a given types is an invalid type and should be ignored.
 	IsInvalidType func(*types.Type) bool
@@ -42,11 +42,11 @@ type DefaultImportTracker struct {
 	PrintImport func(string, string) string
 }

-func NewDefaultImportTracker(local types.Name) DefaultImportTracker {
+func NewDefaultImportTracker(thisPkg string) DefaultImportTracker {
 	return DefaultImportTracker{
 		pathToName: map[string]string{},
 		nameToPath: map[string]string{},
-		local:      local,
+		thisPkg:    thisPkg,
 	}
 }

@@ -56,7 +56,7 @@ func (tracker *DefaultImportTracker) AddTypes(types ...*types.Type) {
 	}
 }
 func (tracker *DefaultImportTracker) AddType(t *types.Type) {
-	if tracker.local.Package == t.Name.Package {
+	if tracker.thisPkg == t.Name.Package {
 		return
 	}

thockin committed:
gengo/v2: don't auto-set the Go header file

Using a hardcoded k/k path is weird - just require it to be set from
outside, if we want it.  It seems to be set everywhere already.

thockin committed:
kube-openapi: don't set gengo GoHeaderFilePath

thockin committed:
codegen client-gen (BROKEN)

TODO: Splats files all over

thockin committed:
WIP: set-gen

thockin committed:
WIP: register-gen

thockin committed:
WIP: import-boss

thockin committed:
WIP: go-to-protobuf

thockin committed:
------------------------------------ LEFT-OFF-HERE

thockin committed:
WIP: gengo/v2: Tweak the imports local-name heuristic

TODO: Re-run generators
TODO: This commit only includes fixes for
TODO:    ./hack/update-codegen.sh prerelease deepcopy defaults conversions openapi

Prefer something longer, which is less ambiguous.  This is all code that
is not often read by humans, so it isn't SO important.

This changed almost all of the generated files, and required some more
discipline in type-refs in openapi generation.

thockin committed:
Convert codegen tools to stop using "vendor"

Workspaces do not (yet?) respect vendoring, so all paths in tools need
to change from `vendor/...` to `staging/src/...`.  This breaks most of
the codegen tools, which are based on gengo.

First change: make codegen stop converting "staging/src" -> "vendor"
in the `find`, which requires golang.sh to get smarter about throwing
the "k8s.io/kubernetes" prefix around.

To fix it all at once would be impossible to review, so this commit
changes each codegen tool's invocation to convert BACK to the old
`vendor...` notation.  This is OK because most of the tools run under
`run-in-gopath.sh` and without GO111MODULE.  As each tool is fixed (in
subsequent commits) these bandaids will be removed.

The eventual results will be:
* no more run-in-gopath
* everything uses GO111MODULE
* no more special-cases for "vendor"
* no more converting "staging/src" to "vendor"

After this commit, update-codegen produces no diffs, but it is still
relying on the existence of vendor.

thockin committed:
------------------- BEGIN CODEGEN HACKING vvvvvvvv

thockin committed:
--------------------- END CODEGEN HACKING ^^^^^^^^

thockin committed:
WIP: Tweak codegen `import`: short form when possible

TODO: Re-run generators
TODO: This commit only includes fixes for
TODO:    ./hack/update-codegen.sh prerelease deepcopy defaults conversions openapi

In particular, do not emit:
   import foo "github/com/bar/foo"
(where the local name is the same as the leaf dir).  Instead emit:
   import "github/com/bar/foo"

This will make future commits cleaner wrt stdlib imports.

This means a bunch of generated files get no-op deltas here.

thockin committed:
WIP: codegen applyconfigurations (BROKEN)

TODO: Emitting weird, broken code

aojea commented:
83 commits 😄

thockin committed:
-------------- BEGIN GO WORKSPACES

pohly commented:
Let me bump this for visibility again: I still think that managing "vendor" more or less like we do now and then injecting its content into an offline build in workspace mode is the best approach - see https://github.com/thockin/kubernetes/pull/13#issuecomment-1201632106 for the implementation and my arguments.

thockin committed:
------ NOTE: REPEAT UPON REBASE ^^^^^^^^^^^^^^^^^^

thockin committed:
------ NOTE: REPEAT UPON REBASE ^^^^^^^^^^^^^^^^^^

dims commented:
If you still need this PR then please rebase, if not, please close the PR

dims commented:
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not

thockin committed:
Fix prerelease codegen log string

thockin committed:
WIP: detailed gengo flow notes

thockin committed:
WIP: add gengo/v2 intra_pkg archetype

thockin committed:
Archetype: gengo owns flag parsing

All the tools seem to do their own flag parsing, but for no obvious
reason.  Change the model so that gengo handles flag parsing.

thockin committed:
WIP

thockin committed:
gengo/v2: Fix namers wrt interfaces and methods

To do this most cleanly, I broke out the inner part of
NameStrategy.Name() into an internal method and changed the Join method
to take a variadic instead of slice.  This change means that
implementations need to be aware.

Then I can recursively call the internal method for interface methods.
So instead of rendering as their Go-syntax name, they render the same as
other Funcs.

I also special-cased `interface{}` -> "Any" and `error` -> "Error".

thockin committed:
Archetype: don't dup namers

thockin committed:
gengo/v2: drop generator.Packages type

thockin committed:
gengo/v2: drop namer.NameSystems type

thockin committed:
Archetype: simpler boilerplate

thockin committed:
Archetype rename DefaultGen

thockin committed:
Archetype rename DefaultPackage

thockin committed:
Archetype clarify SourcePath

thockin committed:
WIP:started-not-much

thockin committed:
WIP: codegen informer-gen

Leaves no diff

thockin committed:
Explicit imports in codegens

Emitted code must add imports for things it depends on.  This worked
automatically for types (via the SnippetWriter and namers) but not for
functions (e.g. fmt.Sprintf()).  Add a new AddImport() method.

Make rawNamer track "this package" as an import.

TODO: Re-rerun generators

thockin committed:
Generated protobuf diffs

All of these are just changes to the descriptor "name" field, e.g.

```diff
@@ -1,5 +1,5 @@
 {
-  "name": "k8s.io/kubernetes/vendor/k8s.io/api/admission/v1/generated.proto",
+  "name": "k8s.io/api/admission/v1/generated.proto",
   "package": "k8s.io.api.admission.v1",
   "dependency": [
     "github.com/gogo/protobuf/gogoproto/gogo.proto",
```

As produced by this program:

```
package main

import (
	"bytes"
	"compress/gzip"
	"encoding/json"
	"fmt"
	"os"

	"io/ioutil"

	proto "github.com/golang/protobuf/proto"
	dpb "github.com/golang/protobuf/protoc-gen-go/descriptor"
)

var before = []byte{
	<data>
}

var after = []byte{
	<data>
}

func main() {
	m := map[string][]byte{
		"before": before,
		"after":  after,
	}
	arg := os.Args[1]
	dump(m[arg])
}

func dump(bytes []byte) {
	fd, err := decodeFileDesc(bytes)
	if err != nil {
		panic(err)
	}
	b, err := json.MarshalIndent(fd, "", "  ")
	if err != nil {
		panic(err)
	}
	fmt.Println(string(b))
}

// decompress does gzip decompression.
func decompress(b []byte) ([]byte, error) {
	r, err := gzip.NewReader(bytes.NewReader(b))
	if err != nil {
		return nil, fmt.Errorf("bad gzipped descriptor: %v", err)
	}
	out, err := ioutil.ReadAll(r)
	if err != nil {
		return nil, fmt.Errorf("bad gzipped descriptor: %v", err)
	}
	return out, nil
}

func decodeFileDesc(enc []byte) (*dpb.FileDescriptorProto, error) {
	raw, err := decompress(enc)
	if err != nil {
		return nil, fmt.Errorf("failed to decompress enc: %v", err)
	}

	fd := new(dpb.FileDescriptorProto)
	if err := proto.Unmarshal(raw, fd); err != nil {
		return nil, fmt.Errorf("bad descriptor: %v", err)
	}
	return fd, nil
}
```

thockin committed:
----------------- gengo/v2 archetypes

thockin committed:
Make golang::binaries_from_targets smarter

and use it from test

Result: `make WHAT=./<something>` works:

```
$ make WHAT="./staging/src/k8s.io/kube-aggregator/pkg/cmd/server"
go version go1.20.4 linux/amd64
+++ [0506 12:58:14] Building go targets for linux/amd64
    ./staging/src/k8s.io/kube-aggregator/pkg/cmd/server (non-static)
```

Also, `make test WHAT=<non-dot-slash-dir>` works like `make` now:

```
$ make all test WHAT=pkg/scheduler/
go version go1.20.4 linux/amd64
+++ [0506 13:17:55] Building go targets for linux/amd64
    ./pkg/scheduler/ (non-static)
go version go1.20.4 linux/amd64
+++ [0506 13:18:03] Setting GOMAXPROCS: 6
+++ [0506 13:18:03] Running tests without code coverage and with -race
ok  	k8s.io/kubernetes/pkg/scheduler	(cached)
```

thockin committed:
Dedup binaries when building

BenTheElder commented:
> /approve cancel

This only cancels your own approval, as opposed to LGTM.
However [WIP] prohibits merging anyhow.

BenTheElder commented:
Whatever hack we come up with for vendor needs to work with `build/run.sh` / dockerized builds as well.

Have we reached out to the Go team about how we might continue with the guarantees of vendor with go.work?


> I found a way around that: when the module download directory contains a .zip file and alongside it a checksum file, then Go trusts that file and compares it against the expected go.sum file. The trick in https://github.com/thockin/kubernetes/pull/13 is to generate a .zip file from the source in vendor and a checksum file that matches the go.sum file.

... is that documented behavior? 🙈

BenTheElder commented on hack/lib/golang.sh:
```diff
@@ -728,7 +728,7 @@ kube::golang::is_instrumented_package() {
 # Echos the path to a dummy test used for coverage information.
```
We should replace this with `go build -cover`
`/assign`

BenTheElder replied:
actually, I'm not sure this is worth it. the results will be incompatible and we're not heavily invested here, and last time this needed fixing with the current approach it was pretty small.

thockin committed:
Simpler and less noisy make-clean

thockin committed:
Archetype: Rename getPackageNamers

thockin committed:
!!!!! DO NOT USE THIS BRANCH !!!!!

thockin committed:
!!!!! IT IS BROKEN !!!!!

thockin committed:
!!!!! USE BRANCH go-workspaces-plus-vendor INSTEAD !!!!!

dims commented:
This work-in-progress PR needs a rebase. please rebase if this PR is still needed for the upcoming release.
